pkg/me/me.go

```diff
@@ -0,0 +1,273 @@
+
+	ret := "Flash partition table:\n"
+	ret += fmt.Sprintf(" Entries: %d\n", h.NumFptEntries)
+	ret += fmt.Sprintf(" HeaderVersion: 0x%x\n", h.HeaderVersion)
+	ret += fmt.Sprintf(" EntryVersion: 0x%x\n", h.EntryVersion)
+	ret += fmt.Sprintf(" HeaderLength: 0x%x\n", h.HeaderLength)
+	ret += fmt.Sprintf(" HeaderChecksum: 0x%x\n", h.HeaderChecksum)
+	ret += fmt.Sprintf(" TicksToAdd: 0x%x\n", h.TicksToAdd)
+	ret += fmt.Sprintf(" TokensToAdd: 0x%x\n", h.TokensToAdd)
+	ret += fmt.Sprintf(" UMASize: 0x%x\n", h.UMASize)
+	ret += fmt.Sprintf(" Flags: 0x%x\n", h.Flags)
+
+	return ret
+}
+
+type FlashPartitionTableHeader struct {
+	Marker             uint32 // Always $FPT
+	NumFptEntries      uint32
+	HeaderVersion      uint8 // Only support 2.0
+	EntryVersion       uint8
+	HeaderLength       uint8 // Usually 0x20
+	HeaderChecksum     uint8
+	TicksToAdd         uint16
+	TokensToAdd        uint16
+	UMASizeOrReserved  uint32
+	FlashLayoutOrFlags uint32
+	// Not Present in ME version 7
+	FitcMajor  uint16
+	FitcMinor  uint16
+	FitcHotfix uint16
+	FitcBuild  uint16
+}
+
+func (h FlashPartitionTableHeader) String() string {
+
+	ret := "Flash partition table:\n"
+	ret += fmt.Sprintf(" Entries: %d\n", h.NumFptEntries)
+	ret += fmt.Sprintf(" HeaderVersion: 0x%x\n", h.HeaderVersion)
+	ret += fmt.Sprintf(" EntryVersion: 0x%x\n", h.EntryVersion)
+	ret += fmt.Sprintf(" HeaderLength: 0x%x\n", h.HeaderLength)
+	ret += fmt.Sprintf(" HeaderChecksum: 0x%x\n", h.HeaderChecksum)
+	ret += fmt.Sprintf(" TicksToAdd: 0x%x\n", h.TicksToAdd)
+	ret += fmt.Sprintf(" TokensToAdd: 0x%x\n", h.TokensToAdd)
+	ret += fmt.Sprintf(" UMASizeOrReserved: 0x%x\n", h.UMASizeOrReserved)
+	ret += fmt.Sprintf(" FlashLayoutOrFlags: 0x%x\n", h.FlashLayoutOrFlags)
+	ret += fmt.Sprintf(" Fitc Version: %d.%d.%d.%d\n", h.FitcMajor, h.FitcMinor, h.FitcHotfix, h.FitcBuild)
+
+	return ret
+}
+
+type FlashPartitionTableEntry struct {
+	Name           [4]uint8
+	Owner          [4]uint8
+	Offset         uint32
+	Length         uint32
+	StartTokens    uint32
+	MaxTokens      uint32
+	ScratchSectors uint32
+	Flags          uint32
+}
+
+func (e FlashPartitionTableEntry) String() string {
+
+	ret := "Flash partition entry:"
+	ret += fmt.Sprintf(" Name: %s\n", []byte{e.Name[0], e.Name[1], e.Name[2], e.Name[3]})
+	ret += fmt.Sprintf(" Owner: %s\n", []byte{e.Owner[0], e.Owner[1], e.Owner[2], e.Owner[3]})
+	ret += fmt.Sprintf(" Offset: 0x%x\n", e.Offset)
+	ret += fmt.Sprintf(" Length: 0x%x\n", e.Length)
+	ret += fmt.Sprintf(" StartTokens: 0x%x\n", e.StartTokens)
+	ret += fmt.Sprintf(" MaxTokens: 0x%x\n", e.MaxTokens)
+	ret += fmt.Sprintf(" ScratchSectors: 0x%x\n", e.ScratchSectors)
+	ret += fmt.Sprintf(" Flags: 0x%x\n", e.Flags)
+
+	if e.Flags>>24 == 0xff {
+		ret += fmt.Sprintf(" Valid: No\n")
+	} else {
+		ret += fmt.Sprintf(" Valid: yes\n")
+	}
+	if e.Flags&1 > 0 {
+		ret += fmt.Sprintf(" Partition: Data\n")
+	} else {
+		ret += fmt.Sprintf(" Partition: Code\n")
+	}
+
+	return ret
+}
```

Comment:
This just a list for every entry? Would be better if the result is a proper table.

Also use strings.Builder

Reply:
fmt.Fprintf(...) can be very helpful here.

Reply:
done